include/miral/miral/decoration_adapter.h

```diff
@@ -0,0 +1,256 @@
+{
+class CursorImages;
+}
+namespace shell
+{
+class SurfaceSpecification;
+namespace decoration
+{
+class Decoration;
+}
+}
+namespace scene
+{
+class Session;
+}
+namespace compositor
+{
+class BufferStream;
+}
+namespace graphics
+{
+class GraphicBufferAllocator;
+class Buffer;
+}
+}
+
+namespace msh = mir::shell;
+namespace msd = msh::decoration;
+namespace ms = mir::scene;
+namespace mc = mir::compositor;
+namespace mg = mir::graphics;
+namespace geometry = mir::geometry;
+
+struct MirEvent;
+
+namespace miral
+{
+namespace decoration
+{
+class WindowState;
+class StaticGeometry;
+class InputContext;
+
+using OnProcessDrag = std::function<void(miral::DeviceEvent device, miral::decoration::InputContext ctx)>;
+using OnProcessEnter = std::function<void(miral::DeviceEvent device, miral::decoration::InputContext ctx)>;
+using OnProcessLeave = std::function<void(miral::decoration::InputContext ctx)>;
+using OnProcessMove = std::function<void(miral::DeviceEvent device, miral::decoration::InputContext ctx)>;
+using OnProcessUp = std::function<void(miral::decoration::InputContext ctx)>;
+using OnProcessDown = std::function<void()>;
+
+using OnWindowAttribChanged = std::function<void(ms::Surface const* window_surface, MirWindowAttrib attrib, int value)>;
+using OnWindowResized =
+    std::function<void(ms::Surface const* window_surface, mir::geometry::Size const& /*window_size*/)>;
+using OnWindowRenamed = std::function<void(ms::Surface const* window_surface, std::string const& name)>;
+using OnWindowStateUpdated = std::function<void(std::shared_ptr<miral::decoration::WindowState>)>;
+using OnScaleChanged = std::function<void(float)>;
+
+struct InputContext
+{
+    InputContext(
+        std::shared_ptr<msh::Shell> const& shell,
+        std::shared_ptr<ms::Session> const& session,
+        std::shared_ptr<ms::Surface> const& window_surface,
+        std::shared_ptr<MirEvent const> const& latest_event,
+        std::shared_ptr<mir::input::CursorImages> const& cursor_images,
+        std::shared_ptr<ms::Surface> const& decoration_surface);
+
+    void request_move() const;
+
+    void request_resize(MirResizeEdge edge) const;
+
+    void set_cursor(std::string const& cursor_image_name) const;
+
+    void request_close() const;
+
+    void request_toggle_maximize() const;
+
+    void request_minimize() const;
+private:
+    std::shared_ptr<msh::Shell> const shell;
+    std::shared_ptr<ms::Session> const session;
+    std::shared_ptr<ms::Surface> const window_surface;
+    std::shared_ptr<const MirEvent> const latest_event;
+    std::shared_ptr<mir::input::CursorImages> const cursor_images;
+    std::shared_ptr<ms::Surface> const decoration_surface;
+};
+
+
+class Renderer
+{
+public:
+    static auto area(geometry::Size const size) -> size_t;
+
+    using Pixel = uint32_t;
+
+    // Can we use even more keywords?
+    static inline auto const buffer_format = mir_pixel_format_argb_8888;
+    static inline auto const bytes_per_pixel = 4;
```

Comment:
This looks like we are only supporting argb-8888 s/w rendering? Might the user want more choice?

Reply:
If these restrictions were part of the example, I would be fine with it. If we're telling shell developers "this is the only way" then I'm not. Our objective is an API we are willing to support indefinitely, that's very different to "while I'm experimenting around".